packages/devtools_server/lib/src/server.dart

```diff
@@ -92,14 +97,117 @@ void serveDevTools({
 
   final server = await shelf.serve(handler, '127.0.0.1', port);
 
+  final devToolsUrl = 'http://${server.address.host}:${server.port}';
+
   printOutput(
-    'Serving DevTools at http://${server.address.host}:${server.port}',
+    'Serving DevTools at $devToolsUrl',
     {
       'method': 'server.started',
       'params': {'host': server.address.host, 'port': server.port, 'pid': pid}
     },
     machineMode: machineMode,
   );
+
+  final Stream<Map<String, dynamic>> _stdinCommandStream = stdin
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .where((String line) => line.startsWith('{') && line.endsWith('}'))
+      .map<Map<String, dynamic>>((String line) {
+    return json.decode(line) as Map<String, dynamic>;
+  });
+
+  // Example input:
+  // {
+  //   "id":0,
+  //   "method":"vm.register",
+  //   "params":{
+  //     "uri":"<vm-service-uri-here>",
+  //   }
+  // }
+  _stdinCommandStream.listen((Map<String, dynamic> json) async {
+    final int id = json['id'];
+    final Map<String, dynamic> params = json['params'];
+
+    if (!params.containsKey('uri')) {
+      printOutput(
+        'Invalid input: $params does not contain the key \'uri\'',
+        {
+          'id': id,
+          'error': 'Invalid input: $params does not contain the key \'uri\'',
+        },
+        machineMode: machineMode,
+      );
+    }
+
+    // json['uri'] should contain a vm service uri.
+    final uri = Uri.parse(params['uri']);
+
+    // Lots of things are considered valid URIs (including empty strings
+    // and single letters) since they can be relative, so we need to do some
+    // extra checks.
+    if (uri != null &&
+        uri.isAbsolute &&
+        (uri.isScheme('ws') ||
+            uri.isScheme('wss') ||
+            uri.isScheme('http') ||
+            uri.isScheme('https')))
+      await registerLaunchDevToolsService(uri, id, devToolsUrl, machineMode);
+  });
+}
+
+Future<void> registerLaunchDevToolsService(
+  Uri uri,
+  int id,
+  String devToolsUrl,
+  bool machineMode,
+) async {
+  try {
+    // Connect to the vm service and register a method to launch DevTools in
+    // chrome.
+    final VmService service = await _connectToVmService(uri);
+
+    service.registerServiceCallback(launchDevToolsService, (request) async {
+      final url = '$devToolsUrl/url=${Uri.encodeComponent(uri.toString())}#';
```

Comment:
Looks like there's a `?` missing here. The URL being built should be the same as the one you'd see in your browser after connecting DevTools to a VM, like `http://localhost:8000/?url=http://localhost:8001/`

It just occurred to me that we're going to lose the additional stuff like `&theme=dark&hide=debugger`, so we may need to fiddle with this a little (possibly passing args in). I think we should get it landed without first though, to avoid complicating this PR more.

Reply:
Add a ? and added a TODO for passing in args.